amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionSettings.scala

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.amqp
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import com.rabbitmq.client.{Address, Connection, ConnectionFactory, ExceptionHandler}
+
+/**
+ * Only for internal implementations
+ */
+sealed trait AmqpConnectionSettings {
+  def getConnection: Connection
+}
+
+final case class ReusableAmqpConnectionSettings(settings: AmqpConnectionSettings) extends AmqpConnectionSettings {
+  private var cachedConnection: Option[Connection] = None
+
+  override def getConnection = cachedConnection match {
+    case Some(connection) => {
+      if (connection.isOpen) connection
+      else {
+        val connection = settings.getConnection
+        cachedConnection = Some(connection)
+        connection
+      }
+    }
+    case None => {
+      val connection = settings.getConnection
+      cachedConnection = Some(connection)
+      connection
+    }
+  }
+
+  def releaseConnection() = cachedConnection match {
+    case Some(connection) => {
+      if (connection.isOpen) connection.close()
+      cachedConnection = None
+    }
+    case None => Unit
+  }
+}
+
+final case class ReusableAmqpConnectionSettingsWithAutomaticRelease(settings: AmqpConnectionSettings)
+    extends AmqpConnectionSettings {
+  private val clients = new AtomicInteger(0)
+  private var cachedConnection: Option[Connection] = None
```

Comment:
This cached connection will be used concurrently by multiple threads so this is not safe. It needs to be thread safe, but not blocking, and we need to ensure only one ongoing connect is happening, so that two concurrent `getConnection` calls does not lead to two separate connections. The atomic counter counter is good thinking though.

One option to work around it would be to just let the user provide a `Connection` instead, and move the responsibility of doing it safely away from here.

Reply:
True.

I'd rather the user providing the settings instead of the connection for the reasons stated on my previous comment (resiliency and reusability).

Would make the methods `getConnection` and `releaseConnection`synchronized on `this` help with this??

Reply:
No, no synchronized in async-land thanks :)

This is doable without resorting to blocking, just a bit tricky.